cmd/dp-import-tracker/main.go

```diff
@@ -107,45 +115,83 @@ func CheckImportJobCompletionState(importAPI *api.ImportAPI, datasetAPI *api.Dat
 		}
 		if instanceFromAPI.State != "completed" && instanceFromAPI.State != "error" {
-			return nil
+			return nil, false
 		}
 		if instanceFromAPI.State == "error" {
 			targetState = instanceFromAPI.State
 		}
 	}
 	// assert: all instances for jobID are marked "completed"/"error", so update import as same
-	if err := importAPI.UpdateImportJobState(jobID, targetState); err != nil {
+	if err := importAPI.UpdateImportJobState(ctx, jobID, targetState); err != nil {
 		log.ErrorC("CheckImportJobCompletionState update", err, log.Data{"jobID": jobID, "last completed instanceID": completedInstanceID})
 	}
-	return nil
+	return nil, false
 }
 
 // manageActiveInstanceEvents handles all updates to trackedInstances in one thread (this is only called once, in its own thread)
 func manageActiveInstanceEvents(
+	ctx context.Context,
 	createInstanceChan chan string,
 	updateInstanceWithObservationsInsertedChan chan insertedObservationsEvent,
 	datasetAPI *api.DatasetAPI,
-	importAPI *api.ImportAPI) {
+	importAPI *api.ImportAPI,
+	instanceLoopDoneChan chan bool) {
+
+	// inform main() when we have stopped processing events
+	defer close(instanceLoopDoneChan)
 
 	trackedInstances := make(trackedInstanceList)
-	if err := trackedInstances.getInstanceList(datasetAPI); err != nil {
+	if err, _ := trackedInstances.getInstanceList(ctx, datasetAPI); err != nil {
 		logFatal("could not obtain initial instance list", err, nil)
 	}
 
-	checkForCompletedInstancesChan := make(chan bool)
+	tickerChan := make(chan bool)
 	go func() {
 		for range time.Tick(checkForCompleteInstancesTick) {
-			checkForCompletedInstancesChan <- true
+			tickerChan <- true
 		}
 	}()
 
-	for {
+	for looping := true; looping; {
 		select {
+		case <-ctx.Done():
+			log.Debug("manageActiveInstanceEvents: loop ending (context done)", nil)
+			looping = false
+		case <-tickerChan:
+			log.Debug("check import Instances", log.Data{"q": trackedInstances})
+			for instanceID := range trackedInstances {
+				stopTracking := false
+				if err, isFatal := trackedInstances.updateInstanceFromDatasetAPI(ctx, datasetAPI, instanceID); err != nil {
+					log.ErrorC("could not update instance", err, log.Data{"instanceID": instanceID, "isFatal": isFatal})
+					stopTracking = isFatal
+				} else if trackedInstances[instanceID].totalObservations > 0 && trackedInstances[instanceID].observationsInsertedCount >= trackedInstances[instanceID].totalObservations {
+					// the above check on totalObservations ensures that we have updated this instance from the Import API (non-default value)
+					// and that inserts have exceeded the expected
+					logData := log.Data{"instanceID": instanceID, "observations": trackedInstances[instanceID].observationsInsertedCount, "jobID": trackedInstances[instanceID].jobID}
+					log.Debug("import instance complete", logData)
+					log.Error(errors.New("TODO check db for actual count(insertedObservations) now that instance appears to be completed (kafka-double-counting?)"), nil)
```

Comment:
Don't really think this is needed? just a TODO comment should be fine, seems noisy to log out every time it happens in the mean time (though arguably, the annoyance could encourage us to change it sooner, but I would cynically assume we'd just get used to ignoring the logs)